arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,755 @@
+  std::ptrdiff_t distep=output.istep(), djstep=output.jstep(), dpstep=output.planestep();
+
+  // Iterate through image and perform cast and any required ops. We could
+  // shorten this code using vil_transform and functors (binding arguments),
+  // but just to ensure they get inlined...
+  if( scale_factor == 1.0 && !rounding_enabled )
+  {
+    vil_convert_cast( input, output );
+  }
+  else if( !rounding_enabled )
+  {
+    const inT* splane = input.top_left_ptr();
+    outT* dplane = output.top_left_ptr();
+    for (unsigned p=0;p<np;++p,splane += spstep,dplane += dpstep)
+    {
+      const inT* srow = splane;
+      outT* drow = dplane;
+      for (unsigned j=0;j<nj;++j,srow += sjstep,drow += djstep)
+      {
+        const inT* spixel = srow;
+        outT* dpixel = drow;
+        for (unsigned i=0;i<ni;++i,spixel+=sistep,dpixel+=distep)
+        {
+          *dpixel = static_cast<outT>(*spixel * scale_factor);
+        }
+      }
+    }
+  }
+  else
+  {
+    const inT* splane = input.top_left_ptr();
+    outT* dplane = output.top_left_ptr();
+    for (unsigned p=0;p<np;++p,splane += spstep,dplane += dpstep)
+    {
+      const inT* srow = splane;
+      outT* drow = dplane;
+      for (unsigned j=0;j<nj;++j,srow += sjstep,drow += djstep)
+      {
+        const inT* spixel = srow;
+        outT* dpixel = drow;
+        for (unsigned i=0;i<ni;++i,spixel+=sistep,dpixel+=distep)
+        {
+          *dpixel = static_cast<outT>(*spixel * scale_factor + 0.5);
+        }
+      }
+    }
+  }
+}
+
+// Cumulative averager
+template <typename PixType>
+cumulative_frame_averager<PixType>
+::cumulative_frame_averager( const bool should_round )
+{
+  this->should_round_ = should_round;
+  this->reset();
+}
+
+template <typename PixType>
+void cumulative_frame_averager<PixType>
+::process_frame( const vil_image_view< PixType >& input,
+                 vil_image_view< PixType >& average )
+{
+  if( this->has_resolution_changed( input ) )
+  {
+    this->reset();
+  }
+
+  // If this is the first frame observed or there was an indicated reset
+  if( this->frame_count_ == 0 )
+  {
+    vil_convert_cast( input, this->last_average_ );
+  }
+  // Standard update case
+  else
+  {
+    // Calculate new average - TODO: Non-exponential cumulative average can be modified
+    // to be more efficient and prevent precision losses by not using math_add_fraction
+    // function. Can also be optimized in the byte case to use integer instead of double
+    // operations, but it's good enough for now.
+    double scale_factor = 1.0 / ( this->frame_count_ + 1 );
+
+    vil_math_add_image_fraction( this->last_average_,
+                                 1.0 - scale_factor,
+                                 input,
+                                 scale_factor );
+  }
+
+  // Copy a completely new image in case we are running in async mode
+  copy_cast_and_scale( this->last_average_,
```

Comment:
I think it is good to return a copy of the image, but this comment may no longer be valid.

Reply:
Are you talking about `async` mode or copying part not being valid? Because I'm not sure what async is in this context but it does look like the copy is being performed correctly.

Reply:
Addressed, thanks.